### ISSUE TITLE
Integration Test Suite Cleanup: Classification, Reclassification, and Fixes

### ISSUE THREAD

**Issue by kpiteira** [kpiteira/ktrdr#268]
## Summary

Our integration test suite has accumulated technical debt through organic growth. A comprehensive analysis reveals **71 failing tests** and **29 skipped tests** that fall into distinct categories requiring different remediation strategies.

This issue documents the complete analysis findings and provides an actionable plan to restore the test suite to a healthy state.

---

## Current State

| Metric | Count |
|--------|-------|
| **Failing** | 71 |
| **Passing** | 341 |
| **Skipped** | 29 |
| **Total** | 441 |

**Test health: 77% passing** (should be 100%)

---

## Root Cause Analysis

The failures stem from several distinct issues:

1. **Unit tests misclassified as integration tests** - Tests that mock everything don't belong in `tests/integration/`
2. **Tests for code in wrong location** - Tests for `training-host-service` code shouldn't be in main test suite
3. **Stale tests after API changes** - Endpoints changed but tests weren't updated
4. **Tests requiring external infrastructure** - DB/Jaeger tests mixed with mockable tests
5. **Lazy skip markers** - Tests skipped with "complex mocking required" excuse instead of being fixed or deleted
6. **Tests for deleted code** - Already addressed: deleted `test_compute_indicator.py` (11 tests for removed CLI command)

---

### Category A: Unit Tests Masquerading as Integration Tests

**Problem:** These tests mock nearly all dependencies and test single classes in isolation. They provide no integration value and inflate the integration test count while testing at the wrong level.

#### `tests/integration/training/test_host_orchestrator.py` (20 tests)

**What it tests:** `HostTrainingOrchestrator` from `training-host-service/orchestrator.py`

**Why it's misclassified:**
- Hacks `sys.path` to import from `training-host-service` directory
- Creates `MockTrainingSession` to avoid host service dependencies  
- Mocks `TrainingPipeline.train_strategy`, `DataManager`, `torch.cuda.is_available`
- Tests internal implementation details (checking source code for sleep operations)

**Evidence from the code:**
```python
# Add training-host-service to path for imports
training_host_service_path = (
    Path(__file__).parent.parent.parent.parent / "training-host-service"
)
if str(training_host_service_path) not in sys.path:
    sys.path.insert(0, str(training_host_service_path))
```

**Test examples:**
- `test_init_accepts_required_parameters` - Unit test
- `test_progress_callback_has_no_sleep_operations` - Inspects source code (definitely unit test)
- `test_no_sleep_in_orchestrator_source` - Inspects source code

**Recommendation:** **MOVE to `training-host-service/tests/unit/`**

These tests belong with the code they test. The sys.path hack is a clear signal they're in the wrong place.

---

#### `tests/integration/training/test_local_orchestrator.py` (5 tests)

**What it tests:** `LocalTrainingOrchestrator` from `ktrdr/api/services/training/local_orchestrator.py`

**Why it's misclassified:**
- Mocks `TrainingPipeline.train_strategy` completely
- Mocks `yaml.safe_load` 
- Mocks `asyncio.to_thread`
- Tests individual methods like `_create_progress_callback`, `_load_strategy_config`

**Evidence from the code:**
```python
with patch(
    "ktrdr.api.services.training.local_orchestrator.TrainingPipeline.train_strategy"
) as mock_train:
    mock_train.return_value = {
        "model_path": "models/test.pth",
        "training_metrics": {},
        "test_metrics": {},
    }
```

**Recommendation:** **MOVE to `tests/unit/api/services/training/`**

---

### Category B: Legitimate Integration Tests (Need Fixes)

These tests verify real interactions between multiple components with minimal/appropriate mocking.

#### `tests/integration/test_worker_e2e.py` (5 tests)

**What it tests:** Integration between `WorkerRegistry` and `WorkerRegistration`

**Why it's legitimate:**
- Creates real `WorkerRegistry` instance
- Creates real `WorkerRegistration` instance  
- Only mocks the HTTP layer (`httpx.AsyncClient.post`) - appropriate for integration tests
- Verifies the complete registration flow works

**Test flow:**
1. Backend initializes WorkerRegistry
2. Worker initializes WorkerRegistration
3. Worker calls `register()` which would POST to backend
4. Mock intercepts HTTP and calls `registry.register_worker()` directly
5. Test verifies registry state

**Why it's failing:** The mock path `httpx.AsyncClient.post` may not match actual usage pattern. Need to investigate how `WorkerRegistration.register()` makes HTTP calls.

**Recommendation:** **FIX** - Update mock path to match actual implementation

---

#### `tests/integration/test_worker_load_balancing.py` (8 tests)

**What it tests:** `WorkerRegistry` behavior with multiple workers

**Why it's legitimate:**
- Creates real `WorkerRegistry` with real workers registered
- Tests actual `select_worker`, `mark_busy`, `mark_available` behavior
- Tests concurrent operations with `asyncio.gather`
- No mocking of core functionality

**Test examples:**
- `test_concurrent_worker_selection` - 30 concurrent selections across 3 workers
- `test_load_balancing_with_busy_workers` - Mark workers busy, verify selection
- `test_round_robin_fairness_over_time` - 100 selections should be evenly distributed

**Why it's failing:** Likely assertion changes - the round-robin implementation may have changed. For example:
```python
# Test expects exactly 25 selections per worker with 100 ops and 4 workers
assert count == 25, f"Worker {worker_id} got {count}, expected 25"
```

**Recommendation:** **FIX** - Verify WorkerRegistry behavior and update assertions

---

#### `tests/integration/test_m6_graceful_shutdown.py` (7 tests)

**What it tests:** Complete graceful shutdown flow for workers

**Why it's legitimate:**
- Tests real `WorkerAPIBase.run_with_graceful_shutdown()` method
- Uses real `TrainingCheckpointState` and `TrainingResumeContext` classes
- Creates actual PyTorch models and saves/loads them
- Tests the integration of shutdown → checkpoint → status update → resume

**Test flow:**
1. Create mock checkpoint service and operations repo (appropriate - external services)
2. Create `GracefulShutdownTestWorker` extending real `WorkerAPIBase`
3. Start operation, trigger `_shutdown_event`
4. Verify checkpoint saved, status updated to CANCELLED
5. Verify resume is possible

**Why it's failing:** Tests expect `GracefulShutdownError` to be raised:
```python
with pytest.raises(GracefulShutdownError):
    await worker.run_with_graceful_shutdown(operation_id, operation())
```
Error: `Failed: DID NOT RAISE <class 'ktrdr.workers.base.GracefulShutdownError'>`

**Recommendation:** **INVESTIGATE** - The `run_with_graceful_shutdown` behavior may have changed. Check if it still raises `GracefulShutdownError` or handles shutdown differently now.

---

### Category C: API Endpoint Tests (Need Fixes)

These are legitimate contract tests that verify API request/response behavior.

#### `tests/integration/api/test_training_endpoints.py` (8 tests)

**What it tests:** Training API endpoints (`POST /api/v1/trainings/start`, etc.)

**Pattern used:**
```python
@pytest.fixture
def client_with_mocked_training(client, mock_training_service):
    from ktrdr.api.endpoints.training import get_training_service
    client.app.dependency_overrides[get_training_service] = lambda: mock_training_service
    yield client
    client.app.dependency_overrides.clear()
```

**Why it's legitimate:** Tests real HTTP routing through FastAPI TestClient with appropriately mocked service layer.

**Why it's failing:** Request payload or response format changed. Test sends:
```python
payload = {
    "symbols": ["AAPL"],
    "timeframes": ["1h"],
    "strategy_name": "rsi_mean_reversion",
    "start_date": "2024-01-01",
    "end_date": "2024-06-01",
}
```
And expects status 200, but endpoint may now require different fields or return different format.

**Recommendation:** **FIX** - Compare test expectations against current API spec

---

#### `tests/integration/api/test_operations_endpoints.py` (5 tests)

**What it tests:** Operations CRUD endpoints

**Why it's failing:** Error response format changed. Tests expect 404 but get 400:
```python
# Test expects 404 for non-existent operation
assert response.status_code == 404
# But API returns 400 with different error format
```

**Recommendation:** **FIX** - Update expected status codes and error format

---

#### `tests/integration/api/test_fuzzy_endpoint.py` (2 tests)

**What it tests:** Fuzzy logic calculation endpoint

**Why it's failing:** Similar error format changes

**Recommendation:** **FIX**

---

#### `tests/integration/api/test_data_endpoints.py` (1 test remaining)

**What it tests:** `GET /api/v1/data/{symbol}/{timeframe}`, `GET /api/v1/symbols`, etc.

**Already fixed:** Deleted 4 tests for removed `POST /api/v1/data/load` endpoint

**Remaining failure:** `test_get_symbols_endpoint_mixed_types` - assertion mismatch

**Recommendation:** **FIX**

---

#### `tests/integration/api/test_data_endpoints_consistency.py` (3 tests remaining)

**What it tests:** DataService delegation patterns (Phase 3 transformation verification)

**Already fixed:** Deleted 5 tests for removed `load_data` endpoint

**Remaining tests:** 
- `test_cached_data_endpoint_no_direct_bypass`
- `test_no_direct_bypass_patterns_remain`
- `test_phase3_requirement_no_direct_bypass`

**Recommendation:** **FIX** - These verify important architectural constraints

---

#### `tests/integration/cli/test_sandbox_commands.py` (5 failing)

**What it tests:** Sandbox CLI registration, help, and commands

**Already fixed:** Added `conftest.py` with `runner` fixture

**Remaining failures:** Assertion mismatches in CLI output

**Recommendation:** **FIX**

---

#### `tests/integration/cli/test_epochs_configuration.py` (1 failing + 3 skipped)

**Failing test:** `test_document_full_flow` 

**Skipped tests (with excuses):**
```python
@pytest.mark.skip(
    reason="Requires complex strategy validation mocking - verified via test_real_strategy_file_epochs_extraction instead"
)
def test_epochs_extracted_from_strategy_yaml(self):
```

**Recommendation:** 
- **FIX** the 1 failing test
- **DELETE** the 3 skipped tests - "complex mocking required" is not a valid reason to keep dead tests

---

### Category E: Tests Requiring Host Services

These tests require real external infrastructure and should be in `tests/host_service/`.

#### `tests/integration/test_db_persistence.py` (6 tests, all skipped)

**Current skip mechanism:**
```python
pytestmark = pytest.mark.skipif(
    not os.getenv("DB_HOST"),
    reason="DB_HOST not configured - database tests skipped"
)
```

**What it tests:** Operation persistence to PostgreSQL

**Recommendation:** **MOVE to `tests/host_service/database/`**

---

#### `tests/host_service/observability/test_jaeger_traces.py` (2 tests)

**Status:** Already moved from `tests/integration/` ✅

---

#### `tests/integration/test_slice_25_infrastructure_validation.py` (1 test)

**Need to check:** What is "slice 25"? Is this test still relevant?

---

#### `tests/integration/training/test_m4_task_4_5_resume_context_integration.py` (2 tests)

**What it tests:** Resume context being passed to training pipeline

**Why it's failing:** Training pipeline signature changes

**Recommendation:** **FIX** if resume feature is still used, **DELETE** if feature removed

---

#### `tests/integration/api/test_dummy_endpoints.py` (1 test)

**What it tests:** OpenAPI spec structure

**Recommendation:** **FIX** - Update expected spec structure

---

### Phase 1: Delete Dead Code (Immediate)
- [x] Delete `test_compute_indicator.py` (done - 11 tests for deleted CLI)
- [x] Delete tests for `POST /api/v1/data/load` (done - 9 tests for deleted endpoint)
- [ ] Delete 3 skipped tests in `test_epochs_configuration.py` with "complex mocking" excuse

### Phase 2: Reclassify Misplaced Tests
- [ ] Move `test_host_orchestrator.py` to `training-host-service/tests/unit/`
- [ ] Move `test_local_orchestrator.py` to `tests/unit/api/services/training/`
- [ ] Move `test_db_persistence.py` to `tests/host_service/database/`

### Phase 3: Fix Legitimate Integration Tests
Priority order (by impact):

1. **Worker tests** (13 tests) - Core distributed architecture
   - [ ] Fix `test_worker_e2e.py` mock path
   - [ ] Fix `test_worker_load_balancing.py` assertions

2. **API tests** (19 tests) - Contract verification
   - [ ] Fix `test_training_endpoints.py`
   - [ ] Fix `test_operations_endpoints.py`
   - [ ] Fix `test_fuzzy_endpoint.py`
   - [ ] Fix remaining `test_data_endpoints.py`
   - [ ] Fix remaining `test_data_endpoints_consistency.py`

3. **CLI tests** (6 tests)
   - [ ] Fix `test_sandbox_commands.py`
   - [ ] Fix `test_epochs_configuration.py`

### Phase 4: Investigate and Decide
- [ ] Investigate `test_m6_graceful_shutdown.py` - Is GracefulShutdownError behavior intentionally changed?
- [ ] Investigate `test_slice_25_infrastructure_validation.py` - Is this still relevant?
- [ ] Investigate `test_m4_task_4_5_resume_context_integration.py` - Is resume feature active?

---

## Expected Outcome

After completing all phases:

| Metric | Before | After |
|--------|--------|-------|
| Failing | 71 | 0 |
| Skipped (lazy) | 3 | 0 |
| Skipped (infrastructure) | 6 | 6 (in host_service) |
| Moved to unit tests | 0 | 25 |
| Moved to host_service | 0 | 6 |
| Deleted (dead code) | 0 | 23 |

**Test health: 77% → 100%**

---

### To Delete
- `tests/integration/cli/test_epochs_configuration.py` - 3 skipped test methods

### To Move to `training-host-service/tests/`
- `tests/integration/training/test_host_orchestrator.py`

### To Move to `tests/unit/`
- `tests/integration/training/test_local_orchestrator.py` → `tests/unit/api/services/training/`

### To Move to `tests/host_service/`
- `tests/integration/test_db_persistence.py` → `tests/host_service/database/`

### To Fix
- `tests/integration/test_worker_e2e.py`
- `tests/integration/test_worker_load_balancing.py`
- `tests/integration/test_m6_graceful_shutdown.py`
- `tests/integration/api/test_training_endpoints.py`
- `tests/integration/api/test_operations_endpoints.py`
- `tests/integration/api/test_fuzzy_endpoint.py`
- `tests/integration/api/test_data_endpoints.py`
- `tests/integration/api/test_data_endpoints_consistency.py`
- `tests/integration/cli/test_sandbox_commands.py`
- `tests/integration/cli/test_epochs_configuration.py`
- `tests/integration/api/test_dummy_endpoints.py`
- `tests/integration/training/test_m4_task_4_5_resume_context_integration.py`
- `tests/integration/test_slice_25_infrastructure_validation.py`

---

## Labels

- `tech-debt`
- `testing`
- `cleanup`
